### PR TITLE
Fix gathering data if keys are not defined

### LINKED ISSUE #64: Unable to use Netbox Testbed Creator after 22.3
I've run into an issue with creating testbed file with `pyats==22.3` and after.
Everything works fine with version `pyats==22.2`.

Netbox version 3.2.3

### version 22.2
`pyats create testbed netbox --output=out --netbox-url=<NETBOX_URL> --user-token=<TOKEN>--verbose  --def_user=<USER> --def_pass=<PASSWORD> --url_filter="platform_id__n=null&has_primary_ip=True"`

```log
Begin retrieving data from netbox...
Starting new HTTPS connection (1): <NETBOX_URL>:443
<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&platform_id__n=null&has_primary_ip=True HTTP/1.1" 200 102346
Starting new HTTPS connection (1): <NETBOX_URL>:443
<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=50&platform_id__n=null HTTP/1.1" 200 109298
Starting new HTTPS connection (1): <NETBOX_URL>:443
<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=100&platform_id__n=null HTTP/1.1" 200 108083
Starting new HTTPS connection (1): <NETBOX_URL>:443
<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=150&platform_id__n=null HTTP/1.1" 200 26860
Starting new HTTPS connection (1): <NETBOX_URL>:443
<NETBOX_URL>:443 "GET /api/virtualization/virtual-machines/?format=json&platform_id__n=null&has_primary_ip=True HTTP/1.1" 200 52
Retrieving associated data for DEV1...
Retrieving associated data for DEV2...
OS type is not valid for DEV3. Skipping...
Testbed file generated:
out
```

### version 22.3 and 22.4
`pyats create testbed netbox --output=out --netbox-url=<NETBOX_URL> --user-token=<TOKEN>--verbose  --def_user=<USER> --def_pass=<PASSWORD> --url_filter="platform_id__n=null&has_primary_ip=True"`

```log
Begin retrieving data from netbox...
Starting new HTTPS connection (1): <NETBOX_URL>:443
https://<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&platform_id__n=null&has_primary_ip=True HTTP/1.1" 200 102346
Starting new HTTPS connection (1): <NETBOX_URL>:443
https://<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=50&platform_id__n=null HTTP/1.1" 200 109298
Starting new HTTPS connection (1): <NETBOX_URL>:443
https://<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=100&platform_id__n=null HTTP/1.1" 200 108083
Starting new HTTPS connection (1): <NETBOX_URL>:443
https://<NETBOX_URL>:443 "GET /api/dcim/devices/?format=json&has_primary_ip=True&limit=50&offset=150&platform_id__n=null HTTP/1.1" 200 26860
Starting new HTTPS connection (1): <NETBOX_URL>:443
https://<NETBOX_URL>:443 "GET /api/virtualization/virtual-machines/?format=json&platform_id__n=null&has_primary_ip=True HTTP/1.1" 200 52
Retrieving associated data for DEV1...
Traceback (most recent call last):
  File "src/pyats/cli/base.py", line 207, in pyats.cli.base.Command.main
  File "src/pyats/cli/base.py", line 337, in pyats.cli.base.CommandWithSubcommands.run
  File "src/genie/cli/commands/create_testbed.py", line 34, in genie.cli.commands.create_testbed.CreateTestbed.run
  File "/home/jankrupa/pyats_newest/venv/lib/python3.8/site-packages/pyats/contrib/creators/creator.py", line 208, in to_testbed_file
    testbed = self._generate()
  File "/home/jankrupa/pyats_newest/venv/lib/python3.8/site-packages/pyats/contrib/creators/netbox.py", line 634, in _generate
    self._get_info(device,self._custom_data_source))
  File "/home/jankrupa/pyats_newest/venv/lib/python3.8/site-packages/pyats/contrib/creators/netbox.py", line 535, in _get_info
    for key in keys:
TypeError: 'NoneType' object is not iterable


'NoneType' object is not iterable
```

Problem with iterating through `None`
https://github.com/CiscoTestAutomation/pyats.contrib/blob/release_22.4/src/pyats/contrib/creators/netbox.py#L535

### PR DESCRIPTION
**Pull request by Kani999**
Close #64

### CODE REVIEW

**Comment by domachad** [comment]
Hi @Kani999, thank you for your contribution. Would it be possible for you to add a unit test that covers this fix?

**Comment by Kani999** [comment]
> Hi @Kani999, thank you for your contribution. Would it be possible for you to add a unit test that covers this fix?

Hello @domachad,
I don't know how to add some tests. It should create some docker instance of Netbox, to create a `Netbox (TextbedCreator)` connection or something like that? 

Because in [pyats.contrib](https://github.com/CiscoTestAutomation/pyats.contrib)/[src](https://github.com/CiscoTestAutomation/pyats.contrib/tree/master/src)/[pyats](https://github.com/CiscoTestAutomation/pyats.contrib/tree/master/src/pyats)/[contrib](https://github.com/CiscoTestAutomation/pyats.contrib/tree/master/src/pyats/contrib)/[creators](https://github.com/CiscoTestAutomation/pyats.contrib/tree/master/src/pyats/contrib/creators)/[tests](https://github.com/CiscoTestAutomation/pyats.contrib/tree/master/src/pyats/contrib/creators/tests)/test_netbox.py file I'll create a test:

```
    def test_generate_testbed(self):
        with self.assertRaises(Exception):
            netbox_testbed_creator = Netbox(user_token="abc", netbox_url="abc", custom_data_source=[])
            netbox_testbed_creator._generate() # This fails, because URL is invalid, so I could not test inner functionality
```

**Comment by Kani999** [comment]
@domachad I've created a "proof of concept" branch which adds the unit test. The branch require `pytest-docker` installed via pip and installed `docker` at the machine. 

There's a comparation of changes:
   - https://github.com/CiscoTestAutomation/pyats.contrib/compare/master...Kani999:fix_netbox_testbed_creator_with_docker_tests?expand=1

**Comment by domachad** [comment]
> @domachad I've created a "proof of concept" branch which adds the unit test. The branch require `pytest-docker` installed via pip and installed `docker` at the machine.
> 
> There's a comparation of changes:
> 
> * https://github.com/CiscoTestAutomation/pyats.contrib/compare/master...Kani999:fix_netbox_testbed_creator_with_docker_tests?expand=1

We avoid adding dependencies whenever possible. I believe you should be able to mock the class and/or methods you might need to call. The point of adding the UT is to have as much code coverage as we can. If anybody makes any changes that might break our code, then we will be able to check that before we merge it.

**Comment by Kani999** [comment]
@domachad  So if I want to call method [_generate()](https://github.com/CiscoTestAutomation/pyats.contrib/pull/65) I have to mock data which is returned by calling multiple API endpoints. 
I'll try to look at it, but I've never written a test like this.

**Comment by Kani999** [comment]
@domachad I've added a unit test which mocks the API calls that are called in `_generate()` method. 
It needs to install package `responses`. I did not find any `requirements_dev.txt` or any development packages in the setup. Where should be the dependency included?

**Review by domachad — CHANGES_REQUESTED** [review]

**Comment by Kani999** [comment]
> @Kani999 Thank you for your contribution. Unit tests are failing, can you please fix that?

@ademz Unit tests are fixed. Can you run the CI?

File: src/pyats/contrib/creators/tests/test_netbox.py [hunk 1]
```diff
@@ -11,5 +15,19 @@ def test_missing_arguments(self):
...
+        responses.add(responses.GET, "http://mocked_url.com" + NETBOX_VMS_API_CALL,
+                      json=NETBOX_VMS_API_RESPONSE, status=200)
+
+        netbox_creator = Netbox(user_token="0123456789abcdef0123456789abcdef01234567",
+                                netbox_url="http://mocked_url.com", custom_data_source=None, def_user="test", def_pass="test")
+
+        testbed = netbox_creator._generate()
+
+        self.assertDictEqual(testbed, NETBOX_GENERATED_TESTBED)
+
```
[the remark this answers is not in the document]
Comment (Kani999) [reply src/pyats/contrib/creators/tests/test_netbox.py]: @domachad I've applied your changes. Thank you

**Review by domachad — APPROVED** [review]

**Pull request merged**